corrected promise-based execution

batrudinych opened the pull request.

robertkowalski commented:
can you describe what it corrects?

batrudinych commented:
> can you describe what it corrects?

sure
* [here](https://github.com/bitfinexcom/bfx-facs-grc/blob/71ea9a9b1ee708b0cab857634f4f0cc551e2ebd2/index.js#L290-L291) we're using `_cb` whilest it might be omitted
* if we're using it with promises but passing the wrong action / args, we should have a rejection
* defaults for `_opts = {}` makes sense if we're using it with promises only. In order to be able to omit it, an `if` statement is added
* `map` method was not supporting promise-based execution

robertkowalski commented:
code lgtm, testing now

robertkowalski reviewed (changes requested):
because this is touching a lot of production services, special care has to be taken. 

in #16 i wrote integration tests that pass on current master and cover the current behaviour.

when rebasing against the PR i get failures when using callbacks:

```
(17:36:03) [robert@tequila-new] ~/bitfinex/bfx-facs-grc (pr/15) $ mocha test/


  RPC integration
    1) request with callback, basic
foo:test:net getIpInfo [] Error: ERR_GRAPE_LOOKUP_EMPTY
    at request (/Users/robert/bitfinex/bfx-facs-grc/node_modules/grenache-nodejs-link/index.js:164:19)
    at Object.cb (/Users/robert/bitfinex/bfx-facs-grc/node_modules/grenache-nodejs-link/node_modules/async/dist/async.js:4617:26)
    at Link.handleReply (/Users/robert/bitfinex/bfx-facs-grc/node_modules/grenache-nodejs-link/index.js:118:9)
    at cbq0.push (/Users/robert/bitfinex/bfx-facs-grc/node_modules/grenache-nodejs-link/index.js:79:12)
    at q.forEach.cb (/Users/robert/bitfinex/bfx-facs-grc/node_modules/cbq/index.js:22:21)
```

suggested steps: integrate #16 and fix the issues. increase test coverage to cover the fixed issues from the PR description.

batrudinych commented:
hey @robertkowalski 
i changed the code and added tests
as `req` and `map` were essentially the same code, I restructured it a bit
the only thing is that `map` will log an error now which it wasn't doing previously
you can see it here: https://github.com/batrudinych/bfx-facs-grc/blob/b82d212b7643458ff799ef535b6334dabe972e67/index.js#L317-L319

robertkowalski commented on test/index.test.js:
```diff
@@ -0,0 +1,147 @@
+  describe('_doReq', () => {
+    describe('callback', () => {
+      it('options defaults', done => {
+        fac._doReq('request', 'rest:util:net', 'getIpInfo', ['8.8.8.8'], (_err, res) => {
```
why not testing `.req` and `.map`, the methods that are used?

batrudinych replied:
because both `req` and `map` call `_doReq`. I can add small tests for `map` and `req` to ensure they're calling `_doReq` with proper params

robertkowalski commented on test/index.test.js:
```diff
@@ -0,0 +1,147 @@
+    fac.stop()
+  })
+
+  describe('_doReq', () => {
```
lets have one file per feature that is tested, else we end up with 500 LOC testfiles where you scroll all day

batrudinych replied:
i'm testing a single function here, which is `_doReq`. Do you want me to split callback / promise tests to different files?

robertkowalski reviewed (changes requested):
nice progress, thanks for documenting the change in behaviour through the additional log, that should not be a problem.

for the tests, they should test the public methods `req` and `map`, so its 100% clear that the code that is used in prod is tested. right now there is no additional value in tests by testing the `_doReq` method directly

robertkowalski reviewed (approved):
tested both with unit tests and with a function in bfx-util-net-js

asked @rbndg for another pair of eyes

batrudinych commented on test/req.test.js:
```diff
@@ -0,0 +1,128 @@
+      }
+      throw new Error('Should have thrown')
+    })
+  })
```
hello @rbndg @robertkowalski, added tests for `ERR_DOUBLE_CB`

however, I failed to add tests for `ERR_TIMEOUT`. Here is the thing, `monitor` returns `ERR_TIMEOUT` [here](https://github.com/bitfinexcom/grenache-nodejs-base/blob/95e11bbe87b4ae3ce225b9c5b04962353de4c919/lib/TransportRPCClient.js#L91-L92) and is called once in 10 seconds.

If I set `{ timeout: 11000 }` when calling `fac.req('rest:util:net',...)`, monitor will check if `now` > start + timeout and won't return an error as now = start + 10000.

At the same time, if I set `{ timeout: 5000 }`, I'll receive `Error: ERR_REQUEST_GENERIC: ESOCKETTIMEDOUT` [here](https://github.com/bitfinexcom/grenache-nodejs-http/blob/d52f979b99f31c94b52fc44825fd26c22d59959e/lib/TransportRPCClient.js#L89).

I might be missing something. Is there a way to receive `ERR_TIMEOUT` at all?

robertkowalski replied:
a manual call to monitor can force it

```js
    it('timeout error', done => {
      fac.req('rest:util:net', 'getTimeoutErr', ['8.8.8.8'], { timeout: 1000 }, (err, res) => {
          // yadda yadda

      })

      fac.peer.monitor()
    }).timeout(7000)
```

batrudinych replied:
alright, I see. I'll use the manual call in tests. Anyway, I have a feeling that as `monitor` is not being called manually usually, the probability to get `ERR_TIMEOUT` in the real world example is close to 0.

rbndg replied:
Interesting, maybe we dont' need that `ERR_TIMEOUT`? @robertkowalski

robertkowalski replied:
`ERR_TIMEOUT` does trigger still, but its very very hard to test programatically. let's omit the test

robertkowalski reviewed (approved):
lgtm

prdn commented on index.js:
```diff
@@ -319,53 +341,18 @@ class Grc extends Base {
+      })
     }
 
-    const peer = service.indexOf('sec:') === 0 ? this.peerSec : this.peer
```
why this was removed?

batrudinych replied:
it's done [in _doReq here](https://github.com/bitfinexcom/bfx-facs-grc/pull/15/files/09ad7687089afdeb12e37411b9f4547aef4648db#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R338)
